

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Quickstart &mdash; pyrefinebio 0.3.0 documentation</title>
  

  
  
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/css/style.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Using the CLI" href="cli.html" />
    <link rel="prev" title="Welcome to pyrefinebio!" href="index.html" /> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          
  <a href="https://www.refine.bio/" class="wy-side-nav-search__back">Back to refine.bio</a>
  

          
            <a href="index.html" class="icon icon-home"> pyrefinebio
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          

        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Quickstart</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#installation">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#setting-up-tokens">Setting up Tokens</a></li>
<li class="toctree-l2"><a class="reference internal" href="#downloading-datasets">Downloading Datasets</a></li>
<li class="toctree-l2"><a class="reference internal" href="#downloading-compendia">Downloading Compendia</a></li>
<li class="toctree-l2"><a class="reference internal" href="#getting-help">Getting Help</a></li>
<li class="toctree-l2"><a class="reference internal" href="#getting-started-with-the-cli">Getting Started with the CLI</a></li>
<li class="toctree-l2"><a class="reference internal" href="#interacting-with-the-api">Interacting with the API</a></li>
<li class="toctree-l2"><a class="reference internal" href="#advanced-dataset-usage">Advanced Dataset Usage</a></li>
<li class="toctree-l2"><a class="reference internal" href="#advanced-token-usage-1">Advanced Token Usage</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">Using the CLI</a><ul>
<li class="toctree-l2"><a class="reference internal" href="cli.html#setting-up-tokens">Setting up Tokens</a></li>
<li class="toctree-l2"><a class="reference internal" href="cli.html#downloading-datasets">Downloading Datasets</a></li>
<li class="toctree-l2"><a class="reference internal" href="cli.html#downloading-compendia">Downloading Compendia</a></li>
<li class="toctree-l2"><a class="reference internal" href="cli.html#getting-information-about-pyrefinebio-classes-and-functions">Getting Information About pyrefinebio Classes and Functions</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="classes.html">Classes</a><ul>
<li class="toctree-l2"><a class="reference internal" href="compendium.html">Compendium</a></li>
<li class="toctree-l2"><a class="reference internal" href="computational_result.html">Computational Result</a></li>
<li class="toctree-l2"><a class="reference internal" href="computed_file.html">Computed File</a></li>
<li class="toctree-l2"><a class="reference internal" href="dataset.html">Dataset</a></li>
<li class="toctree-l2"><a class="reference internal" href="experiment.html">Experiment</a></li>
<li class="toctree-l2"><a class="reference internal" href="institution.html">Institution</a></li>
<li class="toctree-l2"><a class="reference internal" href="organism.html">Organism</a></li>
<li class="toctree-l2"><a class="reference internal" href="original_file.html">Original File</a></li>
<li class="toctree-l2"><a class="reference internal" href="platform.html">Platform</a></li>
<li class="toctree-l2"><a class="reference internal" href="downloader_job.html">Downloader Job</a></li>
<li class="toctree-l2"><a class="reference internal" href="processor_job.html">Processor Job</a></li>
<li class="toctree-l2"><a class="reference internal" href="survey_job.html">Survey Job</a></li>
<li class="toctree-l2"><a class="reference internal" href="processor.html">Processor</a></li>
<li class="toctree-l2"><a class="reference internal" href="qn_target.html">QN Target</a></li>
<li class="toctree-l2"><a class="reference internal" href="sample.html">Sample</a></li>
<li class="toctree-l2"><a class="reference internal" href="token.html">Token</a></li>
<li class="toctree-l2"><a class="reference internal" href="transcriptome_index.html">Transcriptome Index</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="more_stuff.html">More Stuff</a><ul>
<li class="toctree-l2"><a class="reference internal" href="high_level_functions.html">High Level Functions</a></li>
<li class="toctree-l2"><a class="reference internal" href="paginated_list.html">Paginated List</a></li>
<li class="toctree-l2"><a class="reference internal" href="config.html">Config</a></li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">pyrefinebio</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Quickstart</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <a href="https://github.com/AlexsLemonade/refinebio-py/blob/HEAD/docs/quickstart.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <section id="quickstart">
<h1>Quickstart<a class="headerlink" href="#quickstart" title="Permalink to this headline">¶</a></h1>
<section id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<p>pyrefinebio can be installed through PyPI:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ pip install pyrefinebio
</pre></div>
</div>
</section>
<section id="setting-up-tokens">
<span id="quickstart-setting-up-tokens"></span><h2>Setting up Tokens<a class="headerlink" href="#setting-up-tokens" title="Permalink to this headline">¶</a></h2>
<p>Before you start downloading or interacting with the API you should set up a Token.
Tokens will allow you to be able to download Compendia, Computed Files, and Datasets.</p>
<p>Creating and activating a refine.bio Token indicates agreement with refine.bio’s <a class="reference external" href="https://www.refine.bio/terms">Terms of Use</a> and
<a class="reference external" href="https://www.refine.bio/privacy">Privacy Policy</a> so make sure you have read and understood both before you continue.</p>
<p>pyrefinebio provides the function <code class="code docutils literal notranslate"><span class="pre">create_token()</span></code> which can automatically create, activate, and save a Token for you.</p>
<p>By default, <code class="code docutils literal notranslate"><span class="pre">create_token()</span></code> will prompt you before activating and saving the Token it creates.</p>
<p>Alternatively, it has the parameters <code class="code docutils literal notranslate"><span class="pre">agree_to_terms</span></code> and <code class="code docutils literal notranslate"><span class="pre">save_token</span></code> which you can set to bypass the prompts.</p>
<p>Here’s an example of creating, activating, and saving a Token with prompts:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>&gt;&gt;&gt; pyrefinebio.create_token<span class="o">()</span>
Please review the refine.bio Terms of Use: https://www.refine.bio/terms and Privacy Policy: https://www.refine.bio/privacy
Do you understand and accept both documents? <span class="o">(</span>y/N<span class="o">)</span>y
Would you like to save your Token to the Config file <span class="k">for</span> future use? <span class="o">(</span>y/N<span class="o">)</span>y
</pre></div>
</div>
<p>Here’s an example of creating and activating but not saving a Token without prompts:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>&gt;&gt;&gt; pyrefinebio.create_token<span class="o">(</span><span class="nv">agree_to_terms</span><span class="o">=</span>True, <span class="nv">save_token</span><span class="o">=</span>False<span class="o">)</span>
</pre></div>
</div>
<p>Alternatively, you can manually create, activate and save a Token using the Token class.
See <a class="reference internal" href="#advanced-token-usage"><span class="std std-ref">Advanced Token Usage</span></a> for more information.</p>
<p>You can also use the CLI to create, activate and save a Token.
See <a class="reference internal" href="cli.html#using-the-cli-setting-up-tokens"><span class="std std-ref">Setting up Tokens</span></a> for more information.</p>
</section>
<section id="downloading-datasets">
<h2>Downloading Datasets<a class="headerlink" href="#downloading-datasets" title="Permalink to this headline">¶</a></h2>
<p>After you set up and activate a Token you can start creating and downloading Datasets.</p>
<p>See <a class="reference internal" href="#quickstart-setting-up-tokens"><span class="std std-ref">Setting up Tokens</span></a> for a tutorial on setting up a Token.</p>
<p>pyrefinebio provides the function <code class="code docutils literal notranslate"><span class="pre">download_dataset()</span></code> for creating and downloading Datasets.
It will automatically handle every part of the creation and download process for you.
You will receive the Dataset as a zip file.</p>
<p><code class="code docutils literal notranslate"><span class="pre">download_dataset()</span></code> requires that you pass in the parameters <code class="code docutils literal notranslate"><span class="pre">path</span></code>, <code class="code docutils literal notranslate"><span class="pre">email_address</span></code>, and either <code class="code docutils literal notranslate"><span class="pre">dataset_dict</span></code> or <code class="code docutils literal notranslate"><span class="pre">experiments</span></code>.</p>
<ul class="simple">
<li><p><strong>path</strong> <cite>(str)</cite> - The path that the Dataset will be downloaded to. You specify a path to a zip file or a directory. If you pass in a path to a directory, the name of the zip file will be automatically generated in the format <code class="code docutils literal notranslate"><span class="pre">dataset-&lt;dataset_id&gt;.zip</span></code>.</p></li>
<li><p><strong>email_address</strong> <cite>(str)</cite> - The email address that will be notified when the Dataset is finished processing.</p></li>
</ul>
<p>The parameters <code class="code docutils literal notranslate"><span class="pre">experiments</span></code> and <code class="code docutils literal notranslate"><span class="pre">dataset_dict</span></code> both control which Experiments and Samples will be a part of the Dataset.</p>
<ul class="simple">
<li><p><strong>experiments</strong> <cite>(list)</cite> - Should be used when you want to add specific Experiments to your Dataset. All the downloadable samples associated with the Experiments that you pass in will be added to the Dataset.</p></li>
</ul>
<p>The <code class="code docutils literal notranslate"><span class="pre">experiments</span></code> parameter is just a list of Experiment accession codes or pyrefinebio Experiment objects. Here’s an example:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">ex2</span> <span class="o">=</span> <span class="n">pyrefinebio</span><span class="o">.</span><span class="n">Experiment</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;&lt;Experiment 2 Accession Code&gt;&quot;</span><span class="p">)</span>

<span class="n">exs</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;&lt;Experiment 1 Accession Code&gt;&quot;</span><span class="p">,</span> <span class="n">ex2</span><span class="p">]</span>
</pre></div>
</div>
<ul class="simple">
<li><p><strong>dataset_dict</strong> <cite>(dict)</cite> - Should be used when you want to specify specific Samples to be included in the Dataset. However, you can pass in “ALL” instead of specific Sample accession codes to add all downloadable Samples associated with that Experiment to the Dataset.</p></li>
</ul>
<p><cite>dataset_dict</cite> should be in the format:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">dd</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;&lt;Experiment 1 Accession Code&gt;&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="s2">&quot;&lt;Sample 1 Accession Code&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;Sample 2 Accession Code&gt;&quot;</span>
    <span class="p">],</span>
    <span class="s2">&quot;&lt;Experiment 2 Accession Code&gt;&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="s2">&quot;&lt;Sample 3 Accession Code&gt;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&lt;Sample 4 Accession code&gt;&quot;</span>
    <span class="p">],</span>
    <span class="s2">&quot;&lt;Experiment 3 Accession Code&gt;&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="s2">&quot;ALL&quot;</span>
    <span class="p">]</span>
<span class="p">}</span>
</pre></div>
</div>
<p>You can also pass in other optional parameters to alter the Dataset itself and to alter how the download process works.</p>
<ul class="simple">
<li><p><strong>aggregation</strong> <cite>(str)</cite> - Can be used to change how the Dataset is aggregated. The default is “EXPERIMENT”, and the other available choices are “SPECIES” and “ALL”. For more information about Dataset aggregation check out <a class="reference external" href="https://refinebio-docs.readthedocs.io/en/latest/main_text.html?highlight=aggregation#aggregations">Aggregations</a>.</p></li>
<li><p><strong>transformation</strong> <cite>(str)</cite> - Can be used to change the transformation of the Dataset. The default is “NONE”, and the other available choices are “MINMAX” and “STANDARD”. For more information on Dataset transformation check out <a class="reference external" href="https://refinebio-docs.readthedocs.io/en/latest/main_text.html?highlight=quantile#gene-transformations">Gene transformations</a>.</p></li>
<li><p><strong>skip_quantile_normalization</strong> <cite>(bool)</cite> - Can be used to choose whether or not quantile normalization is skipped for RNA-seq Samples. For more information check out <a class="reference external" href="https://refinebio-docs.readthedocs.io/en/latest/main_text.html?highlight=quantile%20normalization#quantile-normalization">Quantile normalization</a>.</p></li>
<li><p><strong>extract</strong> <cite>(bool)</cite> - Can be used to choose whether the downloaded zip file should be automatically extracted. It will automatically extract to the same location that you passed in as <code class="code docutils literal notranslate"><span class="pre">path</span></code>. So if <code class="code docutils literal notranslate"><span class="pre">path</span></code> is a zip file: <code class="code docutils literal notranslate"><span class="pre">./path/to/dataset.zip</span></code> it will be extracted to the dir <code class="code docutils literal notranslate"><span class="pre">./path/to/dataset/</span></code>, if <code class="code docutils literal notranslate"><span class="pre">path</span></code> is a dir: <code class="code docutils literal notranslate"><span class="pre">./path/to/dir/</span></code> it will be extracted to <code class="code docutils literal notranslate"><span class="pre">./path/to/dir/[generated-file-name]/</span></code>. By default, <code class="code docutils literal notranslate"><span class="pre">extract</span></code> is False.</p></li>
<li><p><strong>prompt</strong> <cite>(bool)</cite> - Can be used to choose whether or not you should be prompted before downloading if the Dataset zip file is larger than 1 gigabyte. By default, <code class="code docutils literal notranslate"><span class="pre">prompt</span></code> is True.</p></li>
</ul>
<p>Below is a simple example of downloading a Dataset using <code class="code docutils literal notranslate"><span class="pre">experiments</span></code>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pyrefinebio</span><span class="o">.</span><span class="n">download_dataset</span><span class="p">(</span>
    <span class="s2">&quot;~/path/to/dir/for/dataset/&quot;</span><span class="p">,</span>
    <span class="s2">&quot;foo@bar.com&quot;</span><span class="p">,</span>
    <span class="n">experiments</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;GSE24528&quot;</span><span class="p">,</span> <span class="s2">&quot;GSE30631&quot;</span><span class="p">]</span>
<span class="p">)</span>
</pre></div>
</div>
<p>Below is a simple example of downloading a Dataset using <code class="code docutils literal notranslate"><span class="pre">dataset_dict</span></code>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pyrefinebio</span><span class="o">.</span><span class="n">download_dataset</span><span class="p">(</span>
    <span class="s2">&quot;./path/to/dataset.zip&quot;</span><span class="p">,</span>
    <span class="s2">&quot;foo@bar.com&quot;</span><span class="p">,</span>
    <span class="n">dataset_dict</span><span class="o">=</span><span class="p">{</span>
        <span class="s2">&quot;GSE24528&quot;</span><span class="p">:</span> <span class="p">[</span>
            <span class="s2">&quot;GSM604796&quot;</span><span class="p">,</span>
            <span class="s2">&quot;GSM604797&quot;</span>
        <span class="p">]</span>
    <span class="p">}</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="downloading-compendia">
<h2>Downloading Compendia<a class="headerlink" href="#downloading-compendia" title="Permalink to this headline">¶</a></h2>
<p>You can start downloading Compendia after you set up and activate a Token.</p>
<p>See <a class="reference internal" href="#quickstart-setting-up-tokens"><span class="std std-ref">Setting up Tokens</span></a> for a tutorial on setting up a Token.</p>
<p>pyrefinebio provides the function <code class="code docutils literal notranslate"><span class="pre">download_compendium()</span></code> for downloading Compendium results.
It will automatically search for Compendia based on organisms and download the results.
You will receive the Compendium as a zip file.</p>
<p><cite>download_compendium()</cite> requires that you pass in the parameters <code class="code docutils literal notranslate"><span class="pre">path</span></code> and <code class="code docutils literal notranslate"><span class="pre">organism</span></code>.</p>
<ul class="simple">
<li><p><strong>path</strong> <cite>(str)</cite> - The path that the Dataset will be downloaded to. You specify a path to a zip file or a directory. If you pass in a path to a directory, the name of the zip file will be automatically generated in the format <code class="code docutils literal notranslate"><span class="pre">compendium-&lt;compendium_id&gt;.zip</span></code>.</p></li>
<li><p><strong>organism</strong> <cite>(str)</cite> - The scientific name of the Organism for the Compendium that you want to download.</p></li>
</ul>
<p>You can also pass in other optional parameters to alter the type of Compendium you download.</p>
<ul class="simple">
<li><p><strong>version</strong> <cite>(int)</cite> - The Compendium version. The default is <code class="code docutils literal notranslate"><span class="pre">None</span></code> which will get the latest version.</p></li>
<li><p><strong>quant_sf_only</strong> <cite>(bool)</cite> - Can be used to choose if the Compendium is quantile normalized. Pass in True for RNA-seq Sample Compendium results or False for quantile normalized. By default, <code class="code docutils literal notranslate"><span class="pre">quant_sf_only</span></code> is False. For more information on normalized vs RNA-seq compendia check out <a class="reference external" href="http://docs.refine.bio/en/latest/main_text.html#refine-bio-compendia">refine.bio Compendia</a>.</p></li>
<li><p><strong>extract</strong> <cite>(bool)</cite> - Can be used to choose whether the downloaded zip file should be automatically extracted. It will automatically extract to the same location that you passed in as <code class="code docutils literal notranslate"><span class="pre">path</span></code>. So if <code class="code docutils literal notranslate"><span class="pre">path</span></code> is a zip file: <code class="code docutils literal notranslate"><span class="pre">./path/to/dataset.zip</span></code> it will be extracted to the dir <cite>./path/to/dataset/</cite>, if <code class="code docutils literal notranslate"><span class="pre">path</span></code> is a dir: <code class="code docutils literal notranslate"><span class="pre">./path/to/dir/</span></code> it will be extracted to <code class="code docutils literal notranslate"><span class="pre">./path/to/dir/[generated-file-name]/</span></code>. By default, <code class="code docutils literal notranslate"><span class="pre">extract</span></code> is False.</p></li>
<li><p><strong>prompt</strong> <cite>(bool)</cite> - Can be used to choose whether or not you should be prompted before downloading if the Dataset zip file is larger than 1 gigabyte. By default, <code class="code docutils literal notranslate"><span class="pre">prompt</span></code> is True.</p></li>
</ul>
<p>Below is a simple example of Downloading a Compendium result:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pyrefinebio</span><span class="o">.</span><span class="n">download_compendium</span><span class="p">(</span>
    <span class="s2">&quot;./path/to/compendium.zip&quot;</span><span class="p">,</span>
    <span class="s2">&quot;DANIO_RERIO&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
<p>pyrefinebio also provides the function <code class="code docutils literal notranslate"><span class="pre">download_quantfile_compendium()</span></code> which is equivalent to calling <code class="code docutils literal notranslate"><span class="pre">download_compendium(quant_sf_only=True)</span></code>.</p>
<p>You can use this function when you want to be explicit to future users of your script that you are downloading quantfile Compendium results.</p>
<p>Below is a simple example of Downloading a Compendium result using <code class="code docutils literal notranslate"><span class="pre">download_quantfile_compendium()</span></code>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pyrefinebio</span><span class="o">.</span><span class="n">download_quantfile_compendium</span><span class="p">(</span>
    <span class="s2">&quot;~/path/to/dir/for/compendium/&quot;</span><span class="p">,</span>
    <span class="s2">&quot;HOMO_SAPIENS&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</section>
<section id="getting-help">
<h2>Getting Help<a class="headerlink" href="#getting-help" title="Permalink to this headline">¶</a></h2>
<p>If you are re-reading a script that you wrote and forget what a pyrefinebio function or class does -
or if you just want more information about a pyrefinebio class or function, pyrefinebio has a <code class="code docutils literal notranslate"><span class="pre">help()</span></code> function
that can print out information about all its classes/functions.</p>
<p>The <code class="code docutils literal notranslate"><span class="pre">help()</span></code> function is probably most useful in the REPL.</p>
<p>Here’s an example:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>&gt;&gt;&gt; import pyrefinebio
&gt;&gt;&gt; pyrefinebio.help<span class="o">(</span><span class="s2">&quot;download_dataset&quot;</span><span class="o">)</span>
</pre></div>
</div>
<p>This will print out information about the <code class="code docutils literal notranslate"><span class="pre">download_dataset()</span></code> function.</p>
<p>This is what the output looks like:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>download_dataset

    Automatically constructs a Dataset, processes it, waits <span class="k">for</span> it
    to finish processing, <span class="k">then</span> downloads it to the path specified.

    Returns:
        Dataset

    Parameters:
        path <span class="o">(</span>str<span class="o">)</span>: the path that the Dataset should be downloaded to

        email_address <span class="o">(</span>str<span class="o">)</span>: the email address that will be contacted with info
                                related to the dataset

        dataset_dict <span class="o">(</span>dict<span class="o">)</span>: a fully formed Dataset <span class="sb">`</span>data<span class="sb">`</span> attribute <span class="k">in</span> the form:
                                <span class="o">{</span>
                                    <span class="s2">&quot;Experiment&quot;</span>: <span class="o">[</span>
                                        <span class="s2">&quot;Sample&quot;</span>,
                                        <span class="s2">&quot;Sample&quot;</span>
                                    <span class="o">]</span>
                                <span class="o">}</span>
                                use this parameter <span class="k">if</span> you want to specify specific Samples
                                <span class="k">for</span> your dataset
                                each part of the dict can be a pyrefinebio object or an accession
                                code as a string

        experiments <span class="o">(</span>list<span class="o">)</span>: a list of Experiments that should be added to the dataset
                            use this parameter <span class="k">if</span> you only care about the Experiments - all
                            available samples related to each Experiment will be added
                            the list can contain Experiment objects or accession codes as strings

        aggregation <span class="o">(</span>str<span class="o">)</span>: how the Dataset should be aggregated - by <span class="sb">`</span>EXPERIMENT<span class="sb">`</span> or by <span class="sb">`</span>SPECIES<span class="sb">`</span>

        transformation <span class="o">(</span>str<span class="o">)</span>: the transformation <span class="k">for</span> the dataset - <span class="sb">`</span>NONE<span class="sb">`</span>, <span class="sb">`</span>MINMAX<span class="sb">`</span>, or <span class="sb">`</span>STANDARD<span class="sb">`</span>

        skip_quantile_normalization <span class="o">(</span>bool<span class="o">)</span>: control whether or not the dataset should skip quantile
                                            normalization <span class="k">for</span> RNA-seq Samples

        extract <span class="o">(</span>bool<span class="o">)</span>: <span class="k">if</span> true, the downloaded zip file will be automatically extracted

        prompt <span class="o">(</span>bool<span class="o">)</span>: <span class="k">if</span> true, will prompt before downloading files bigger than 1GB
</pre></div>
</div>
<p>You can also get information on class methods by passing in <code class="code docutils literal notranslate"><span class="pre">class.method</span></code> to the help function.</p>
<p>Here’s an example:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>&gt;&gt;&gt; import pyrefinebio
&gt;&gt;&gt; pyrefinebio.help<span class="o">(</span><span class="s2">&quot;Sample.search&quot;</span><span class="o">)</span>
</pre></div>
</div>
</section>
<section id="getting-started-with-the-cli">
<h2>Getting Started with the CLI<a class="headerlink" href="#getting-started-with-the-cli" title="Permalink to this headline">¶</a></h2>
<p>pyrefinebio provides a CLI that exposes the <code class="code docutils literal notranslate"><span class="pre">create_token()</span></code>, <code class="code docutils literal notranslate"><span class="pre">download_dataset()</span></code>, <code class="code docutils literal notranslate"><span class="pre">download_compendium()</span></code>, <code class="code docutils literal notranslate"><span class="pre">download_quantfile_compendium()</span></code>, and <code class="code docutils literal notranslate"><span class="pre">help()</span></code> functions.</p>
<p>Each function has its own command: <code class="code docutils literal notranslate"><span class="pre">create-token</span></code>, <code class="code docutils literal notranslate"><span class="pre">download-dataset</span></code>, <code class="code docutils literal notranslate"><span class="pre">download-compendium</span></code>, <code class="code docutils literal notranslate"><span class="pre">download-quantfile-compendium</span></code>, and <code class="code docutils literal notranslate"><span class="pre">describe</span></code>, respectively.</p>
<p>To use the CLI just type <code class="code docutils literal notranslate"><span class="pre">refinebio</span> <span class="pre">COMMAND</span></code> into a shell.</p>
<p>Each command has the option <code class="code docutils literal notranslate"><span class="pre">--help</span></code> which will print out usage information and descriptions for every available option for that command.</p>
<p>If you want usage examples and a more in depth look at each CLI command you can also check out <a class="reference internal" href="cli.html#using-the-cli"><span class="std std-ref">Using the CLI</span></a>.</p>
<p>Here’s an example of downloading a Dataset using the CLI:</p>
<div class="highlight-shell notranslate"><div class="highlight"><pre><span></span>$ refinebio download-dataset --path <span class="s2">&quot;./path/to/dataset.zip&quot;</span> --email-address <span class="s2">&quot;foo@bar.com&quot;</span> --dataset-dict <span class="s1">&#39;{&quot;GSE74410&quot;: [&quot;ALL&quot;]}&#39;</span>
</pre></div>
</div>
</section>
<section id="interacting-with-the-api">
<h2>Interacting with the API<a class="headerlink" href="#interacting-with-the-api" title="Permalink to this headline">¶</a></h2>
<p>You can use pyrefinebio to interact with all endpoints of the refine.bio API.</p>
<p>Each endpoint is its own pyrefinebio class.</p>
<p>Most classes provide the <code class="code docutils literal notranslate"><span class="pre">get()</span></code> and <code class="code docutils literal notranslate"><span class="pre">search()</span></code> methods.</p>
<p>Use <code class="code docutils literal notranslate"><span class="pre">get()</span></code> to get one refine.bio API object based on its identifying property.</p>
<p>Here’s an example with a refine.bio Sample:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">sample</span> <span class="o">=</span> <span class="n">pyrefinebio</span><span class="o">.</span><span class="n">Sample</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;GSM604797&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Use <code class="code docutils literal notranslate"><span class="pre">search()</span></code> to look for refine.bio API objects based on filters.</p>
<p><code class="code docutils literal notranslate"><span class="pre">search()</span></code> will return a PaginatedList which can be indexed and iterated through like a regular python list.
For more information checkout the <a class="reference internal" href="paginated_list.html#paginatedlist"><span class="std std-ref">Paginated List</span></a> documentation.</p>
<p>Here’s an example of searching for a refine.bio Sample:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">samples</span> <span class="o">=</span> <span class="n">pyrefinebio</span><span class="o">.</span><span class="n">Sample</span><span class="o">.</span><span class="n">search</span><span class="p">(</span>
    <span class="n">organism</span><span class="o">=</span><span class="s2">&quot;HOMO_SAPIENS&quot;</span><span class="p">,</span>
    <span class="n">is_processed</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="n">sample</span> <span class="o">=</span> <span class="n">samples</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
</pre></div>
</div>
<p>Other classes provide additional methods like <code class="code docutils literal notranslate"><span class="pre">download()</span></code> or <code class="code docutils literal notranslate"><span class="pre">extract()</span></code>.</p>
<p>For more information checkout the <a class="reference internal" href="classes.html#classes"><span class="std std-ref">Classes</span></a> documentation.</p>
</section>
<section id="advanced-dataset-usage">
<h2>Advanced Dataset Usage<a class="headerlink" href="#advanced-dataset-usage" title="Permalink to this headline">¶</a></h2>
<p>pyrefinebio offers the <code class="code docutils literal notranslate"><span class="pre">download_dataset()</span></code> function which takes care of creating the Dataset, processing it, waiting for it to finish,
downloading it, and optionally extracting it - all automatically.</p>
<p>If you only want to do a certain part of this process, however, you can do each step manually as well.</p>
<p>To manually create a Dataset, first create an object of the Dataset class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">dataset</span> <span class="o">=</span> <span class="n">pyrefinebio</span><span class="o">.</span><span class="n">Dataset</span><span class="p">()</span>
</pre></div>
</div>
<p>Before you are able to process the Dataset you must add an email address and Experiments/Samples to it.</p>
<p>You can add the email address when creating the Dataset object by passing it into the constructor.
Or you can just set the attribute after creating the object.</p>
<p>Here’s an example of both:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># using the constructor</span>
<span class="n">dataset</span> <span class="o">=</span> <span class="n">pyrefinebio</span><span class="o">.</span><span class="n">Dataset</span><span class="p">(</span><span class="n">email_address</span><span class="o">=</span><span class="s2">&quot;foo@bar.com&quot;</span><span class="p">)</span>

<span class="c1"># setting the attribute</span>
<span class="n">dataset</span> <span class="o">=</span> <span class="n">pyrefinebio</span><span class="o">.</span><span class="n">Dataset</span><span class="p">()</span>
<span class="n">dataset</span><span class="o">.</span><span class="n">email_address</span> <span class="o">=</span> <span class="s2">&quot;foo@bar.com&quot;</span>
</pre></div>
</div>
<p>You can add Experiments and Samples to your dataset by manually setting its <code class="code docutils literal notranslate"><span class="pre">data</span></code> attribute or by using the <code class="code docutils literal notranslate"><span class="pre">add_samples()</span></code> method.</p>
<p>Here’s an example of manually setting its <code class="code docutils literal notranslate"><span class="pre">data</span></code>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">dataset</span><span class="o">.</span><span class="n">data</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s2">&quot;GSE24528&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="s2">&quot;GSM604796&quot;</span><span class="p">,</span>
        <span class="s2">&quot;GSM604797&quot;</span>
    <span class="p">]</span>
<span class="p">}</span>
</pre></div>
</div>
<p>Here’s an example of using the <code class="code docutils literal notranslate"><span class="pre">add_samples()</span></code> method.
Notice that <code class="code docutils literal notranslate"><span class="pre">add_samples()</span></code> can take pyrefinebio objects as arguments as well as accession codes.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">s</span> <span class="o">=</span> <span class="n">pyrefinebio</span><span class="o">.</span><span class="n">Sample</span><span class="o">.</span><span class="n">search</span><span class="p">(</span>
    <span class="n">experiment_accession_code</span><span class="o">=</span><span class="s2">&quot;GSE60783&quot;</span><span class="p">,</span>
    <span class="n">is_processed</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="n">dataset</span><span class="o">.</span><span class="n">add_samples</span><span class="p">(</span>
    <span class="s2">&quot;GSE60783&quot;</span><span class="p">,</span>
    <span class="n">samples</span><span class="o">=</span><span class="p">[</span><span class="n">s</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">s</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
<span class="p">)</span>
</pre></div>
</div>
<p>Once you have set the email address and added Experiments/Samples to you Dataset,
you can then start processing the Dataset using the <code class="code docutils literal notranslate"><span class="pre">process()</span></code> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">dataset</span><span class="o">.</span><span class="n">process</span><span class="p">()</span>
</pre></div>
</div>
<p>You can check if the Dataset has finished at any time by calling the <code class="code docutils literal notranslate"><span class="pre">check()</span></code> method on it:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="n">dataset</span><span class="o">.</span><span class="n">check</span><span class="p">():</span>
    <span class="c1"># do something...</span>
</pre></div>
</div>
<p>Then once the Dataset has finished processing, you can download it using the <code class="code docutils literal notranslate"><span class="pre">download()</span></code> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">dataset</span><span class="o">.</span><span class="n">download</span><span class="p">(</span><span class="s2">&quot;./path/to/dataset.zip&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Once the Dataset has been downloaded, you can extract the downloaded zip file with the <code class="code docutils literal notranslate"><span class="pre">extract()</span></code> method:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">dataset</span><span class="o">.</span><span class="n">extract</span><span class="p">()</span>
</pre></div>
</div>
<p>For more information checkout the <a class="reference internal" href="dataset.html#dataset"><span class="std std-ref">Dataset</span></a> documentation.</p>
</section>
<section id="advanced-token-usage-1">
<span id="advanced-token-usage"></span><h2>Advanced Token Usage<a class="headerlink" href="#advanced-token-usage-1" title="Permalink to this headline">¶</a></h2>
<p>To create a token, make an object of the Token class:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pyrefinebio</span>

<span class="n">token</span> <span class="o">=</span> <span class="n">pyrefinebio</span><span class="o">.</span><span class="n">Token</span><span class="p">()</span>
</pre></div>
</div>
<p>Then, to activate your token, call <code class="code docutils literal notranslate"><span class="pre">agree_to_terms_and_conditions()</span></code> on it:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">token</span><span class="o">.</span><span class="n">agree_to_terms_and_conditions</span><span class="p">()</span>
</pre></div>
</div>
<p>Now, your Token is fully set up and will be used in any API requests made during the execution of this script.</p>
<p>If you want to use the same Token in future scripts, you can save it to the Config file:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">token</span><span class="o">.</span><span class="n">save_token</span><span class="p">()</span>
</pre></div>
</div>
<p>Now, the token will be automatically loaded every time you import pyrefinebio.</p>
<p>If you want to get access to the token that is saved to the config file in the future, you can load it:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">token</span> <span class="o">=</span> <span class="n">pyrefinebio</span><span class="o">.</span><span class="n">Token</span><span class="o">.</span><span class="n">load_token</span><span class="p">()</span>
</pre></div>
</div>
<p>For more information check out the <a class="reference internal" href="token.html#token"><span class="std std-ref">Token</span></a> class documentation and <a class="reference internal" href="config.html#config"><span class="std std-ref">Config</span></a>.</p>
</section>
</section>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="cli.html" class="btn btn-neutral float-right" title="Using the CLI" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="index.html" class="btn btn-neutral" title="Welcome to pyrefinebio!" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, CCDL.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0.3.0',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  <script type="text/javascript" src="_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
   
  <script type="text/javascript" src="_static/js/sticky-sidebar.min.js"></script>
  <script type="text/javascript">
    (function() {
      var stickySidebar = new StickySidebar('.wy-side-scroll');
      document.addEventListener('click', function(e) {
        if (!event.target.matches('.wy-side-scroll a span')) return;
        stickySidebar.updateSticky();
      }, true)
    })();
  </script>


</body>
</html>